manticore/core/memory.py

```diff
@@ -1071,6 +1135,211 @@ def _try_get_solutions(self, address, size, access, max_solutions=0x1000, force=
+        :param filename: the pathname to the file to map.
+        :param offset: the contents of a file mapping are initialized using C{size} bytes starting
+                      at offset C{offset} in the file C{filename}.
+        :return: the starting address where the file was mapped.
+        :rtype: int
+        :raises error:
+                   - 'Address shall be concrete' if C{addr} is not an integer number.
+                   - 'Address too big' if C{addr} goes beyond the limit of the memory.
+                   - 'Map already used' if the piece of memory starting in C{addr} and with length C{size} isn't free.
+        '''
+        # If addr is NULL, the system determines where to allocate the region.
+        assert addr is None or isinstance(addr, int), 'Address shall be concrete'
+        assert addr < self.memory_size, 'Address too big'
+        assert size > 0
+
+        map = AnonMap(addr, size, perms)
+        self._add(map)
+
+        # address is rounded down to the nearest multiple of the allocation granularity
+        if addr is not None:
+            addr = self._floor(addr)
+
+        # size value is rounded up to the next page boundary
+        size = self._ceil(size)
+
+        with open(filename, 'rb') as f:
+            fdata = f.read()  # fdata is a bytes now
+
+        # this worked
+        fdata = fdata[offset:]
+        fdata = fdata.ljust(size, b'\0')
+
+        for i in range(size):
+            Memory.write(self, addr + i, chr(fdata[i]), force=True)
+
+        logger.debug('New file-memory map @%x size:%x', addr, size)
+
+        return addr
+
+    def _deref_can_succeed(self, mapping, address, size):
+        if not issymbolic(address):
+            return address >= mapping.start and address + size < mapping.end
+        else:
+            constraint = Operators.AND(address >= mapping.start, address + size < mapping.end)
+            return solver.can_be_true(self.constraints, constraint)
+
+    def _import_concrete_memory(self, from_addr, to_addr):
+        """
+        for each address in this range need to read from concrete and write to symbolic
+        it's possible that there will be invalid/unmapped addresses in this range. need to skip to next map if so
+        also need to mark all of these addresses as now in the symbolic store
+
+        :param int from_addr:
+        :param int to_addr:
+        :return:
+        """
+        logger.debug("Importing concrete memory: {:x} - {:x} ({} bytes)".format(from_addr, to_addr, to_addr - from_addr))
+
+        for m in self.maps:
+            span = interval_intersection(m.start, m.end, from_addr, to_addr)
+            if span is None:
+                continue
+
+            start, stop = span
+
+            for addr in range(start, stop):
+                if addr in self.backed_by_symbolic_store:
+                    continue
+
+                self.backing_array[addr] = Memory.read(self, addr, 1)[0]
```

Comment:
Wouldn't it be better to read until stop or until it is backed by symbolic store?